### ISSUE TITLE
Action click push notifocation

### ISSUE THREAD

**Issue by ChanhTran255** [evollu/react-native-fcm#648]
I can't get action click push notification. I want get action click because after click will open url . Thank you.
## My code.
` componentDidMount() {
//console.log('wrapper did mount')
NetInfo.addEventListener('change', this._handleConnectionInfoChange );
NetInfo.fetch().done (
(connectionInfo) => { this._handleConnectionInfoChange(connectionInfo); }
);

    //Config Push Notification
    if(Platform.OS == 'ios') {
        FCM.requestPermissions()
        .then(e => {})
        .catch(() => {})
    }
    FCM.getFCMToken().then(token => {
        // console.log('token', token)
        if(token && token != undefined) {                
            this._saveDeviceToken(token)
        }
    })
    this.refreshTokenListener = FCM.on(FCMEvent.RefreshToken, (token) => {  
        // console.log('token refresh', token)          
        if(token && token != undefined) {
            this._saveDeviceToken(token)
        }                
    })
    FCM.getInitialNotification().then(notification => {
        if(notification && notification.opened_from_tray) {
            this._openUrl(this.state.url);
        }
    });
    this.notificationListener = FCM.on(FCMEvent.Notification, (notification) => {
        if(notification)
        {            
            const url_open = notification.url;
            this.setState({
                url: url_open ? url_open : false
            })
            if(notification.local_notification) { 
                if(notification.opened_from_tray) {
                    this._openUrl(this.state.url);
                }
                return;
            }
            if(notification.opened_from_tray) {
                return;
            }
            const title = notification.fcm.title;
            FCM.presentLocalNotification({
                priority: 'high',
                title: title ? title : Params.Name,
                body: notification.fcm.body,
                big_text: notification.fcm.body,
                icon: 'pushicon',
                large_icon: 'joinmunch',
                url : url_open,
                lights: true,
                show_in_foreground: true,
                local: true
            })  
            FCM.cancelAllLocalNotifications()
        }
    })
}`

**Comment by evollu** [comment]
this.setState({
            url: url_open ? url_open : false
        })
sets states in next event loop. that might be your issue?

Does the callback get called?

**Comment by ChanhTran255** [comment]
When get notification i save url. I want get action click notification.

**Comment by evollu** [comment]
sorry I don't quite understand what you want

**Comment by ChanhTran255** [comment]
When you get a notification, when you touch it, it will open the app for you. I want to catch the event when you touch it. Previously, I could use get action 

 `FCM.getInitialNotification().then(notification => {
            if(notification && notification.opened_from_tray) {
                this._openUrl(this.state.url);
            }
        });`
to click. But current i can't.

**Comment by evollu** [comment]
why currently you can't? someone opened an issue that notification.opened_from_tray is undefined in getInitialNotification

**Comment by ChanhTran255** [comment]
Previously after touch notification it will go to getInitialNotification. Current i don't see.

**Comment by ChanhTran255** [comment]
### My Code

componentDidMount() {
        //console.log('wrapper did mount')
        NetInfo.addEventListener('change', this._handleConnectionInfoChange );
        NetInfo.fetch().done (
           (connectionInfo) => { this._handleConnectionInfoChange(connectionInfo); }
        );
        
        //Config Push Notification
        if(Platform.OS == 'ios') {
            FCM.requestPermissions()
            .then(e => {})
            .catch(() => {})
        }
        
        FCM.getFCMToken().then(token => {
            if(token && token != undefined) {                
                this._saveDeviceToken(token)
            }
        })

        this.refreshTokenListener = FCM.on(FCMEvent.RefreshToken, (token) => {  
            if(token && token != undefined) {
                this._saveDeviceToken(token)
            }                
        })

        FCM.getInitialNotification().then(notification => {
            if(notification && notification.opened_from_tray) {
                this._openUrl(this.state.url);
            }
        });
        this.notificationListener = FCM.on(FCMEvent.Notification, (notification) => {
            if(notification)
            {            
                const url_open = notification.url;
                this.setState({
                    url: url_open ? url_open : false
                })
                if(notification.local_notification) { 
                    if(notification.opened_from_tray) {
                        this._openUrl(this.state.url);
                    }
                    return;
                }
                if(notification.opened_from_tray) {
                    return;
                }
                const title = notification.fcm.title;
                FCM.presentLocalNotification({
                    priority: 'high',
                    title: title ? title : Params.Name,
                    body: notification.fcm.body,
                    big_text: notification.fcm.body,
                    icon: 'pushicon',
                    large_icon: 'joinmunch',
                    url : url_open,
                    lights: true,
                    show_in_foreground: true,
                    local: true
                })  
                FCM.cancelAllLocalNotifications()
            }
        })
    }

**Comment by evollu** [comment]
I haven't changed any behavior recently. don't you see it in ` FCM.on(FCMEvent.Notification`?

**Comment by ChanhTran255** [comment]
yes. when click i don't see any one event in FCM.on(FCMEvent.Notification

**Comment by evollu** [comment]
so you are saying when you tap on notification you don't see any callback in javascript?
How about moving the listeners out of component lifecycle?

**Comment by ChanhTran255** [comment]
Previously my code same now. It still working when tab on notification but current it not wroking.

**Comment by dashracer** [comment]
Hi @evollu, I'm having the same problem. When notification received, I can log it from notification listener. When I click the notification,it show FCM without any other data


![fcm android on click](https://user-images.githubusercontent.com/2896774/33308117-cb35fec4-d44b-11e7-9ef3-009922363be2.png)

Any idea to solve this issue?

**Comment by evollu** [comment]
@dashracer let me check

**Comment by evollu** [comment]
wait. what OS are you using?

**Comment by dashracer** [comment]
@evollu Those screenshot from android

**Comment by evollu** [comment]
I don't have such issue with the example project in android
![screenshot_1511974589](https://user-images.githubusercontent.com/9213224/33411526-2ee7b256-d554-11e7-96da-09ae28bea92c.png)

**Comment by DikshaDeep** [comment]
I am also facing the same issue as @ChanhTran255.
I didn't get the data when i click on notification.

**Comment by rohit75** [comment]
I also want an event when notification is clicked. I want to navigate to particular screen when notification is pressed. Any Solution?
